text/0003-templates.md

```diff
@@ -0,0 +1,447 @@
+- It becomes possible to reference multiple templates the package's author does not control with less ambiguity. In an inheritance-based mechanism, multiple inheritance would be necessary for such functionality. However, multiple inheritance can lead to [diamond-shaped problems](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) and less clarity as to what the final value of the field should be.
+
+A few of the problems above can be mitigated by rendering the template package itself before publishing, but this introduces other problems. Toolkit templates install dependencies in the referencing package as `devDependencies`, which isn't desirable in other templates.
+
+The existing design is inspired by the concept of ["_composition over inheritance_"](https://en.wikipedia.org/wiki/Composition_over_inheritance), which provides reusability without defining difficult to change relationships between packages.
+
+### Specifiers
+
+Templates can be specified through:
+
+1. The `name@version` syntax to refer to an external template fetched from a registry.
+2. The `name@workspace:` syntax to refer to a template that's also a workspace package.
+3. The `file:<path>` syntax to refer to an on-disk path.
+
+### Shared Workspace Configuration
+
+It can be tedious to set up the same template configurations for each package in a pnpm workspace. The `workspaceTemplateConfigs` option can be specified at the root `package.json` to simplify.
+
+```json5
+// packages.json
+{
+  "pnpm": {
+    "workspaceTemplateConfigs": {
+      // All packages in the workspace should use this configuration.
+      "default": {
+        "authoring": [
+          "@organization/authoring-metadata@0.1.0",
+          "@team/authoring-metadata@0.1.0"
+        ],
+        "catalog": ["@example/frontend-catalog@workspace:"],
+      },
+
+      // Applies to all packages in the workspace tagged as a "library"
+      "library": {
+        "scripts": ["@example/library-scripts@workspace:"]
+      },
+
+      // Other packages may be tagged as an "application". For this example,
+      // applications have a different set of scripts.
+      "application": {
+        "scripts": ["@example/application-scripts@workspace:"]
+      }
+
+    },
+  }
+}
+```
+
+A special `workspaceConfigs` key can be specified to determine which workspace template configs apply to a given package.
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@example/foo",
+  "pnpm": {
+    "templates": {
+      "workspaceConfigs": ["default", "library"]
+    }
+  }
+}
+```
+
+The example above uses the `default` tag. There is no special treatment for this tag name; it is not applied to all workspace packages implicitly. From a design standpoint, all `package.json` files should specify `pnpm.templates` to communicate to pnpm and other tooling that the file alone is not enough to render its contents.
```

Comment:
Not realizing a `package.json` is "templated" is something I'm concerned about and @rafaelliu also called out earlier. https://github.com/pnpm/rfcs/pull/3#issuecomment-1494868523

> I feel this is a broader issue of adapting to changes tho. Using templates there may still be users editing package.json and getting unexpected results - just change "not realizing there's there's a co-located meta-manifest file" with "not realizing there's a template being used".